### PR TITLE
(ruby) Create .install package and add streams

### LINKED ISSUE #1783: (ruby) Many versions of Ruby have no corresponding version on Chocolatey
## Expected Behavior

Major versions of Ruby increment the second number in the version rather than the first, so, for example, 2.5.8-1 is backwards-compatible with 2.5.3-1, but not 2.4.7-1. There can still be bugs and incompatibilities to resolve in the previous major version, though, so there are still 2.n.x releases after 2.n+1.0-1 is released. One might expect that those releases are available through chocolatey if `--version` is passed.

## Current Behavior

Instead, once 2.n+1.0-1 is released, any further 2.n.x-y releases are ignored by the package generation script. For example, Ruby 2.5.9-1 was released in April 2020, but the latest 2.5.x on Chocolatey is 2.5.3-101 (with the 01 suffix being a fix version from Chocolatey rather than part of the Ruby version) from 2018.

## Possible Solution

* #1298, so there'd be a `ruby2.5` package and a `ruby2.6` package and a `ruby2.7` package etc. plus one `ruby` metapackage that just installed the latest. This has been previously rejected.
* Change the package generation script so it'll pick up minor and patch releases for older major versions, and generate and publish packages for them.

## Context

Because Ruby native gems rely on MSYS2, and MSYS2 has a rolling release model, sometimes a change to MSYS2 breaks Ruby (even if the MSYS2 version's pinned in Chocolatey). Because MSYS2 removed some old versions of automake a week ago with https://github.com/msys2/MSYS2-packages/commit/de632449e0562c3621497448f646046d88a26596, Ruby 2.5.3-101 from Chocolatey doesn't work properly (`ridk install 3` fails). Ruby 2.5.8-1 removed the dependency on these automake MSYS2 packages, but it isn't available through Chocolatey, so if you've got Ruby 2.5 code, you'd need to either stop using the Chocolatey package to install Ruby, or port the Ruby code to a newer version.

Locally, I've worked around this by creating a 2.5.9-1 version of the public package by swapping out the version numbers and files - the package scripts themselves are compatible with the missing versions, it's just the package generation script not creating the missing package versions.

### PR DESCRIPTION
**Pull request by TheCakeIsNaOH**
## Description

This PR does two things.
First, it adds support for AU streams, so updates for old minor versions (2.x, 3.x) get packaged. Versions lower than `2.1` mostly do not have x64 builds, and so are not included.
Second, it creates ruby.install, and moves ruby to a virtual package that depends on ruby.

## Motivation and Context

Fixes #1783 
This issue is requesting that previous minor versions get packaged.

Part of #1757
This issue is for migration of the ruby.portable package to this repository, which entails creating the ruby.install package and moving ruby to a virtual package.

## How Has this Been Tested?

The AU streams has been tested by enabling `NoCheckChocoVersion` and then running `update.ps1`, and checking that all the minor versions (version 2.1 or higher) have packages created. 

Those packages were then installed (side by side) and then uninstalled in the test env to ensure that they work and can all be installed side by side. 

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Migrated package (a package has been migrated from another repository)

## Checklist:

- [x] My code follows the code style of this repository.
- [x] My change requires a change to documentation (this usually means the notes in the description of a package).
- [x] I have updated the documentation accordingly (this usually means the notes in the description of a package).
- [x] All files are up to date with the latest [Contributing Guidelines](https://github.com/chocolatey-community/chocolatey-packages/blob/master/CONTRIBUTING.md)
- [x] The added/modified package passed install/uninstall in the chocolatey test environment.
- [x] The changes only affect a single package (not including meta package).

### CODE REVIEW

File: automatic/ruby/README.md [hunk 1]
```diff
@@ -2,16 +2,8 @@
...
-- `/InstallDir` - Ruby installation directory, by default `c:\tools\RubyXY` where XY are major and minor version parts.
-- `/NoPath`     - Do not add ruby bin folder to machine PATH.
-
-Example: `choco install ruby --params "/NoPath /InstallDir:C:\your\install\path"`
-
 ## Notes
 
-- To install ruby development kit ruby installer provides `ridk` command. It provides easy way to install msys2 via `ridk install 1`, however, the installation is interactive. To accomplish unattended install, use [msys2](https://chocolatey.org/packages/msys2) package.
+- This package depends on the the latest major version of the ruby.install package.
+- This package do not support package parameters, please install the package [ruby.install](https://chocolatey.org/packages/ruby.install) directly when using package parameters.
```
Comment (chrisdjali-wrld3d) [inline automatic/ruby/README.md:8]: Should be *does not*. Same thing in `ruby.nuspec`.

File: automatic/ruby.install/tools/chocolateyInstall.ps1 [hunk 2]
```diff
@@ -12,7 +12,7 @@ if ( !$pp['NoPath'] ) { $tasks += 'modpath'  }
 Write-Host "Ruby is going to be installed in '$installDir'"
 
 $packageArgs = @{
-  packageName    = 'ruby'
+  packageName    = 'ruby.install'
   fileType       = 'exe'
   file           = Get-Item "$toolsDir\*_x32.exe"
   file64         = Get-Item "$toolsDir\*_x64.exe"
```
Comment (AdmiringWorm) [inline automatic/ruby.install/tools/chocolateyInstall.ps1:18]: I know you did not change any lines here, but could you update these two lines to say:

```suggestion
  file           = "$toolsDir\"
  file64         = "$toolsDir\
```

File: automatic/ruby.install/update.ps1 [hunk 3]
```diff
@@ -0,0 +1,64 @@
+import-module au
+
+$releases = 'https://rubyinstaller.org/downloads/archives/'
+
+function global:au_SearchReplace {
+   @{
+        ".\tools\chocolateyInstall.ps1" = @{
+            "(?i)(^\s*packageName\s*=\s*)('.*')"  = "`$1'$($Latest.PackageName)'"
+            "(?i)(^\s*fileType\s*=\s*)('.*')"     = "`$1'$($Latest.FileType)'"
+        }
```
Comment (AdmiringWorm) [inline automatic/ruby.install/update.ps1:10]: In addition to the requested change for the files in the `chocolateyInstall.ps1` script, please add the replacements:

```suggestion
          "(?i)(^\s*file\s*=\s*`"[$]toolsPath\\).*" = "`${1}$($Latest.FileName32)`""
          "(?i)(^\s*file64\s*=\s*`"[$]toolsPath\\).*" = "`${1}$($Latest.FileName64)`""
        }
```

File: automatic/ruby.install/update.ps1 [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+          "(?i)(\s+x32:).*"            = "`${1} $($Latest.URL32)"
+          "(?i)(\s+x64:).*"            = "`${1} $($Latest.URL64)"
+          "(?i)(checksum32:).*"        = "`${1} $($Latest.Checksum32)"
+          "(?i)(checksum64:).*"        = "`${1} $($Latest.Checksum64)"
+          "(?i)(Get-RemoteChecksum).*" = "`${1} $($Latest.URL64)"
+        }
+    }
+}
+
+function global:au_BeforeUpdate { Get-RemoteFiles -Purge }
```
Comment (AdmiringWorm) [inline automatic/ruby.install/update.ps1:22]: This cmdlet should be located before the cmdlet `global:au_SearchReplace`

File: automatic/ruby/README.md [hunk 5]
```diff
@@ -2,16 +2,8 @@
...
-
-- `/InstallDir` - Ruby installation directory, by default `c:\tools\RubyXY` where XY are major and minor version parts.
-- `/NoPath`     - Do not add ruby bin folder to machine PATH.
-
-Example: `choco install ruby --params "/NoPath /InstallDir:C:\your\install\path"`
-
 ## Notes
 
-- To install ruby development kit ruby installer provides `ridk` command. It provides easy way to install msys2 via `ridk install 1`, however, the installation is interactive. To accomplish unattended install, use [msys2](https://chocolatey.org/packages/msys2) package.
+- This package depends on the the latest major version of the ruby.install package.
```
Comment (AdmiringWorm) [inline automatic/ruby/README.md:7]: Do NOT make the description different in the Readme for the `ruby` package, than what it is in the `ruby.install` package.

The Readmes in those two packages should be identical (with the exception of the top link to the package).

File: automatic/ruby/README.md (hunk 1 above)
Comment (AdmiringWorm) [inline automatic/ruby/README.md:8]: > This package do not support package parameters

That is not actually true, meta-packages that only differ in a `.install`, `.portable`, `.tool` and `.commandline` suffixes (bit unsure about the last two) support package parameter if the dependent package uses them.

If a package has other differences, then package parameters isn't supported (think the `python` package that depends on `python3`).

File: automatic/ruby/ruby.nuspec [hunk 6]
```diff
@@ -4,24 +4,16 @@
   <metadata>
     <id>ruby</id>
     <version>3.1.0.1</version>
-    <title>Ruby</title>
+    <title>Ruby (Install)</title>
```
Comment (AdmiringWorm) [inline automatic/ruby/ruby.nuspec:7]: This is not the Install edition of `ruby`, it should only be named `Ruby` for the meta-package.

File: automatic/ruby/ruby.nuspec [hunk 7]
```diff
@@ -34,13 +26,9 @@ Example: `choco install ruby --params "/NoPath /InstallDir:C:\your\install\path"
...
-      <dependency id="chocolatey-core.extension" version="1.3.3" />
-      <dependency id="chocolatey" version="0.10.5" />
+        <dependency id="ruby.install" version="[3.1.0.1]" />
     </dependencies>
   </metadata>
-  <files>
-    <file src="tools\**" target="tools" />
-    <file src="legal\**" target="legal" />
-  </files>
+  <files></files>
```
Comment (AdmiringWorm) [inline automatic/ruby/ruby.nuspec:32]: To follow what is done for other packages.

```suggestion
  <files />
```

**Review by AdmiringWorm — CHANGES_REQUESTED** [review]
Quite a few things that need to be fixed up, and in addition to those can you please fix up the formatting for the PowerShell and Nuspec files?

Indentation should be equal to 2 spaces, no trailing spaces at the end of files and a newline at the end of each file.

See the `.editorconfig` file: https://github.com/chocolatey-community/chocolatey-packages/blob/master/.editorconfig

File: automatic/ruby/README.md (hunk 1 above)
Comment (TheCakeIsNaOH) [reply automatic/ruby/README.md:8]: > That is not actually true, meta-packages that only differ in a .install, .portable, .tool and .commandline suffixes (bit   unsure about the last two) support package parameter if the dependent package uses them.


I thought that it required `--applypackageparameterstodependencies`? So if you do `choco install <package> --params="/switch"`, and `<package>` depends on `<package>.install`, then the parameters do not by default get applied to `<package>install`.

https://github.com/chocolatey/choco/blob/develop/src/chocolatey/infrastructure.app/services/PowershellService.cs#L420-L425

Comment (TheCakeIsNaOH) [reply automatic/ruby/README.md]: Fixed.

**Comment by TheCakeIsNaOH** [comment]
@AdmiringWorm I think i've updated the formatting, I opened it up in vscode with the editorconfig extension, hopefully it's formatter fixed the issues.

There is one question relating to the package parameters with dependencies, see the revew comment.

File: automatic/ruby/README.md (hunk 1 above)
Comment (AdmiringWorm) [reply automatic/ruby/README.md]: Not for pure metapackages (those that either differs in `.install` or `.portable`).

If it had been named differently than the install package, then yes you would have to use that argument.

File: automatic/ruby.install/README.md [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+## Package Parameters
+
+- `/InstallDir` - Ruby installation directory, by default `c:\tools\RubyXY` where XY are major and minor version parts.
+- `/NoPath`     - Do not add ruby bin folder to machine PATH.
+
+Example: `choco install ruby --params "/NoPath /InstallDir:C:\your\install\path"`
+
+## Notes
+
+- To install ruby development kit ruby installer provides `ridk` command. It provides easy way to install msys2 via `ridk install 1`, however, the installation is interactive. To accomplish unattended install, use [msys2](https://chocolatey.org/packages/msys2) package.
```
[the remark this answers is not in the document]
Comment (AdmiringWorm) [reply automatic/ruby.install/README.md:16]: @TheCakeIsNaOH why did you mark this as resolved, when no changes to the file have been made?

**Review by AdmiringWorm — CHANGES_REQUESTED** [review]

File: automatic/ruby/README.md (hunk 5 above)
Comment (TheCakeIsNaOH) [reply automatic/ruby/README.md]: Fixed

File: automatic/ruby/README.md (hunk 1 above)
Comment (TheCakeIsNaOH) [reply automatic/ruby/README.md]: Ok, fixed. I'll need to dig into the Chocolatey source at some point to figure out where this happens.

File: automatic/ruby/ruby.nuspec [hunk 9]
```diff
@@ -1,46 +1,35 @@
 <?xml version="1.0" encoding="utf-8"?>
 <!-- Do not remove this test for UTF-8: if “Ω” doesn’t appear as greek uppercase omega letter enclosed in quotation marks, you should use an editor that supports UTF-8, not this one. -->
-<package xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
+<package xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xmlns:xsd="http://www.w3.org/2001/XMLSchema">
```
[the remark this answers is not in the document]
Comment (TheCakeIsNaOH) [reply automatic/ruby/ruby.nuspec]: Done.

File: automatic/ruby.install/ruby.install.nuspec [hunk 10]
```diff
@@ -0,0 +1,42 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Do not remove this test for UTF-8: if “Ω” doesn’t appear as greek uppercase omega letter enclosed in quotation marks, you should use an editor that supports UTF-8, not this one. -->
+<package xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xmlns:xsd="http://www.w3.org/2001/XMLSchema">
```
[the remark this answers is not in the document]
Comment (TheCakeIsNaOH) [reply automatic/ruby.install/ruby.install.nuspec]: Done

**Comment by AdmiringWorm** [comment]
@TheCakeIsNaOH your changes have been merged, thanks for your contribution 👍

**Pull request merged**